File: examples/ruby/customer/lib/customer_logic.rb
# frozen_string_literal: true

require 'time'
require_relative 'command_validation_error'

# Pure business logic for customer aggregate.
# No gRPC dependencies - can be tested in isolation.
module CustomerLogic
  CustomerState = Struct.new(:name, :email, :loyalty_points, :lifetime_points, keyword_init: true) do
    def initialize(name: '', email: '', loyalty_points: 0, lifetime_points: 0)
      super
    end

    def exists?
      !name.empty?
    end
  end

  class << self
    def rebuild_state(event_book)
      state = CustomerState.new

      return state if event_book.nil? || event_book.pages.empty?

      # Start from snapshot if present
      if event_book.snapshot&.state
        snap = event_book.snapshot.state
        if snap.type_url.end_with?('CustomerState')
          snap_state = Examples::CustomerState.decode(snap.value)
          state = CustomerState.new(
            name: snap_state.name,
            email: snap_state.email,
            loyalty_points: snap_state.loyalty_points,
            lifetime_points: snap_state.lifetime_points
          )
        end
      end

      # Apply events
      event_book.pages.each do |page|
        next unless page.event

        state = apply_event(state, page.event)
      end

      state
    end

    def handle_create_customer(state, name, email)
      raise CommandValidationError.failed_precondition('Customer already exists') if state.exists?
      raise CommandValidationError.invalid_argument('Customer name is required') if name.nil? || name.empty?
      raise CommandValidationError.invalid_argument('Customer email is required') if email.nil? || email.empty?

      Examples::CustomerCreated.new(
        name: name,
        email: email,
        created_at: now_timestamp
      )
    end

    def handle_add_loyalty_points(state, points, reason)
      raise CommandValidationError.failed_precondition('Customer does not exist') unless state.exists?
      raise CommandValidationError.invalid_argument('Points must be positive') if points <= 0

      new_balance = state.loyalty_points + points

      Examples::LoyaltyPointsAdded.new(
        points: points,
        new_balance: new_balance,
        reason: reason || ''
      )
    end

    def handle_redeem_loyalty_points(state, points, redemption_type)
      raise CommandValidationError.failed_precondition('Customer does not exist') unless state.exists?
      raise CommandValidationError.invalid_argument('Points must be positive') if points <= 0
      if points > state.loyalty_points
        raise CommandValidationError.failed_precondition(
          "Insufficient points: have #{state.loyalty_points}, need #{points}"
        )
      end

      new_balance = state.loyalty_points - points

      Examples::LoyaltyPointsRedeemed.new(
        points: points,
        new_balance: new_balance,
        redemption_type: redemption_type || ''
      )
    end

    private

    def apply_event(state, event_any)
      type_url = event_any.type_url

      if type_url.end_with?('CustomerCreated')
        event = Examples::CustomerCreated.decode(event_any.value)
        CustomerState.new(
          name: event.name,
          email: event.email,
          loyalty_points: state.loyalty_points,
          lifetime_points: state.lifetime_points
        )
      elsif type_url.end_with?('LoyaltyPointsAdded')
        event = Examples::LoyaltyPointsAdded.decode(event_any.value)
        CustomerState.new(
          name: state.name,
          email: state.email,
          loyalty_points: event.new_balance,
          lifetime_points: state.lifetime_points + event.points
        )
      elsif type_url.end_with?('LoyaltyPointsRedeemed')
        event = Examples::LoyaltyPointsRedeemed.decode(event_any.value)
        CustomerState.new(
          name: state.name,
          email: state.email,
          loyalty_points: event.new_balance,
          lifetime_points: state.lifetime_points
        )
      else
        state
      end
    end

    def now_timestamp
      Google::Protobuf::Timestamp.new(seconds: Time.now.to_i)
    end
  end
end

File: examples/typescript/customer/src/customer-logic.ts
/**
 * Pure business logic for customer aggregate.
 * No gRPC dependencies - can be tested in isolation.
 */

import {
  CommandValidationError,
  StatusCode,
} from './command-validation-error.js';
import {
  CustomerState,
  emptyCustomerState,
  customerExists,
} from './customer-state.js';

export interface CustomerCreatedEvent {
  name: string;
  email: string;
  createdAt: { seconds: number; nanos: number };
}

export interface LoyaltyPointsAddedEvent {
  points: number;
  newBalance: number;
  reason: string;
}

export interface LoyaltyPointsRedeemedEvent {
  points: number;
  newBalance: number;
  redemptionType: string;
}

export type CustomerEvent =
  | { type: 'CustomerCreated'; data: CustomerCreatedEvent }
  | { type: 'LoyaltyPointsAdded'; data: LoyaltyPointsAddedEvent }
  | { type: 'LoyaltyPointsRedeemed'; data: LoyaltyPointsRedeemedEvent };

export interface EventPage {
  typeUrl: string;
  data: any;
}

export interface EventBook {
  pages: EventPage[];
  snapshot?: {
    state?: any;
  };
}

/**
 * Customer business logic operations.
 */
export class CustomerLogic {
  /**
   * Rebuilds customer state from an event history.
   */
  rebuildState(eventBook: EventBook | null | undefined): CustomerState {
    let state = emptyCustomerState();

    if (!eventBook?.pages?.length) {
      return state;
    }

    // Start from snapshot if present
    if (eventBook.snapshot?.state) {
      const snapState = eventBook.snapshot.state;
      state = {
        name: snapState.name || '',
        email: snapState.email || '',
        loyaltyPoints: snapState.loyaltyPoints || 0,
        lifetimePoints: snapState.lifetimePoints || 0,
      };
    }

    // Apply events
    for (const page of eventBook.pages) {
      state = this.applyEvent(state, page);
    }

    return state;
  }

  private applyEvent(state: CustomerState, page: EventPage): CustomerState {
    const typeUrl = page.typeUrl || '';

    if (typeUrl.endsWith('CustomerCreated')) {
      return {
        ...state,
        name: page.data.name,
        email: page.data.email,
      };
    } else if (typeUrl.endsWith('LoyaltyPointsAdded')) {
      return {
        ...state,
        loyaltyPoints: page.data.newBalance,
        lifetimePoints: state.lifetimePoints + page.data.points,
      };
    } else if (typeUrl.endsWith('LoyaltyPointsRedeemed')) {
      return {
        ...state,
        loyaltyPoints: page.data.newBalance,
      };
    }

    return state;
  }

  /**
   * Handles a CreateCustomer command.
   */
  handleCreateCustomer(
    state: CustomerState,
    name: string,
    email: string
  ): CustomerCreatedEvent {
    if (customerExists(state)) {
      throw CommandValidationError.failedPrecondition('Customer already exists');
    }

    if (!name) {
      throw CommandValidationError.invalidArgument('Customer name is required');
    }
    if (!email) {
      throw CommandValidationError.invalidArgument('Customer email is required');
    }

    return {
      name,
      email,
      createdAt: this.now(),
    };
  }

  /**
   * Handles an AddLoyaltyPoints command.
   */
  handleAddLoyaltyPoints(
    state: CustomerState,
    points: number,
    reason: string
  ): LoyaltyPointsAddedEvent {
    if (!customerExists(state)) {
      throw CommandValidationError.failedPrecondition('Customer does not exist');
    }

    if (points <= 0) {
      throw CommandValidationError.invalidArgument('Points must be positive');
    }

    const newBalance = state.loyaltyPoints + points;

    return {
      points,
      newBalance,
      reason: reason || '',
    };
  }

  /**
   * Handles a RedeemLoyaltyPoints command.
   */
  handleRedeemLoyaltyPoints(
    state: CustomerState,
    points: number,
    redemptionType: string
  ): LoyaltyPointsRedeemedEvent {
    if (!customerExists(state)) {
      throw CommandValidationError.failedPrecondition('Customer does not exist');
    }

    if (points <= 0) {
      throw CommandValidationError.invalidArgument('Points must be positive');
    }

    if (points > state.loyaltyPoints) {
      throw CommandValidationError.failedPrecondition(
        `Insufficient points: have ${state.loyaltyPoints}, need ${points}`
      );
    }

    const newBalance = state.loyaltyPoints - points;

    return {
      points,
      newBalance,
      redemptionType: redemptionType || '',
    };
  }

  private now(): { seconds: number; nanos: number } {
    return { seconds: Math.floor(Date.now() / 1000), nanos: 0 };
  }
}
